### PR TITLE
Feature #2349 upgrade instructions for deprecated MET config files

### LINKED ISSUE #2349: Documentation: Add upgrade instructions for removing user wrapped MET config files
PR #2347 removed support for the deprecated environment variables that were previously set in wrapped MET config files. Most use cases have been updated to no longer use user-defined wrapped MET config files in favor of using the wrapped MET config files provided with the METplus wrappers and settings all configurations via a METplus config file. However, some users may still be using their own out-of-date wrapped MET config files that still reference old environment variables that will no longer be set starting in v6.0.0.

## Describe the Task ##
Provide instructions to help users update their METplus use cases to remove their dependency on their own wrapped MET config files. This should include:
* How to compare user-defined wrapped MET config files to defaults in MET to determine what settings need to be added to METplus config file
* Advice on how to find the equivalent METplus config setting (Python Wrappers chapter of the METplus User's Guide)
* Advice on how to set MET config settings that are not yet supported as METplus config variables (MET config overrides section of User's Guide)

Consider creating a utility script to parse MET config files and determine what needs to be updated.

### Time Estimate ###
~3 days

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED CYCLE ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Task Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
## Change Summary

- Added upgrade instructions to release notes
- Cleaned up error logs for deprecated env vars so it is easier to update using upgrade instructions
- Updated MET Configuration sections in Python Wrappers chapter to:
  - Put wrapped MET config file in expandable block
  - Changed table that shows relationship between METplus and MET configs to use headers so we can link to specific items
- Added support for GRID_STAT_CAT_THRESH to set cat_thresh in GridStat

## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

* Review changes to MET Configuration sections in Python Wrappers chapter: https://metplus.readthedocs.io/en/feature_2349_doc_update_met_config/Users_Guide/wrappers.html#grid-stat-met-conf
* Review upgrade instructions: https://metplus.readthedocs.io/en/feature_2349_doc_update_met_config/Users_Guide/release-notes.html#metplus-wrappers-upgrade-instructions
* Review error log formatting for use case run that requires updates: seneca:/d1/projects/METplus/pull_requests/METplus-6.0.0/pr_2386/metplus.log.20231020201849

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[No]**</br>

- [X] Please complete this pull request review by **10/27/23**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [X] Review the source issue metadata (required labels, projects, and milestone).
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: docs/Users_Guide/release-notes.rst
Comment (j-opatz) [inline docs/Users_Guide/release-notes.rst]: This sentence isn't as understandable as I'd hope for users. Maybe something along the lines of
"After a list of all deprecated environment variables that were found, users can find the path to the files to compare in the final error message" or something of that sort

File: docs/Users_Guide/release-notes.rst
Comment (j-opatz) [inline docs/Users_Guide/release-notes.rst]: In documentation, it's best practice to spell out the whole word. Let's change diff to difference here.
Below you use the command line **diff**, which is OK

File: docs/Users_Guide/release-notes.rst
Comment (j-opatz) [inline docs/Users_Guide/release-notes.rst:178]: Can this line be altered slightly? Maybe to 

Lines that begin with the **<** character are from the first file passed to **diff** (i.e. the user-defined wrapped MET config file).

or 

Lines that begin with the **<** character are from the first file passed to **diff**; in this case, the user-defined wrapped MET config file.

The same logic should apply to the next sentence.

File: docs/Users_Guide/release-notes.rst
Comment (j-opatz) [inline docs/Users_Guide/release-notes.rst]: Unfortunately, it looks like RTD is merging the dashes into one line. Maybe a space between them will work?

Comment (jprestop) [reply docs/Users_Guide/release-notes.rst]: Instead of bolding items like ">" and "---", you could instead consider putting them in back ticks.  e.g. `>` and `---`

File: docs/Users_Guide/release-notes.rst
Comment (j-opatz) [inline docs/Users_Guide/release-notes.rst]: Can we get a link to the top level of the Python Wrappers chapter here?

**Review by j-opatz — APPROVED** [review]
While I suggested some syntax changes, nothing was necessary before the approval of this PR.

However, given the complexity of this instruction guide I strongly recommend running through this with the rest of the METplus team at a future meeting before officially deploying this. That said I am approving this PR.

File: docs/Users_Guide/release-notes.rst
Comment (jprestop) [inline docs/Users_Guide/release-notes.rst]: @georgemccabe What do you think about adding MET_INSTALL_DIR to the beginning of share/met/GridStatConfig_default?  I wasn't sure if the formatting should be ${MET_INSTALL_DIR}/share/met/GridStatConfig_default or {ENV[MET_INSTALL_DIR]}/share/met/GridStatConfig_default or simply MET_INSTALL_DIR/share/met/GridStatConfig_default, which is why I did not directly suggest a change.

File: docs/Users_Guide/release-notes.rst
Comment (jprestop) [inline docs/Users_Guide/release-notes.rst]: Using the "./" here may be misleading, since the instructions do not say to change directories to the METplus installation location.  If a user is not familiar with "./" and is blindly copying and pasting the path, this command may not work.

File: docs/Users_Guide/release-notes.rst
Comment (georgemccabe) [reply docs/Users_Guide/release-notes.rst]: There is no environment variable called MET_INSTALL_DIR that would be set for the user, but there is a METplus config variable with that name, so I think it may be misleading to reference it here. I do see what you are saying about making it more clear how to find the path. I suggest we either change it to:

`<MET_INSTALLATION_LOCATION>/share/met/GridStatConfig`

or

`met-X.Y.Z/share/met/GridStatConfig`

or

`/path/to/met-X.Y.Z/share/met/GridStatConfig`

What do you think?

**Comment by jprestop** [comment]
@georgemccabe On this page:

https://metplus.readthedocs.io/en/feature_2349_doc_update_met_config/Users_Guide/wrappers.html#grid-stat-met-conf

with regard to formatting, I suggest putting the titles of the tables (i.e. "METplus Config(s) | MET Config File") in bold.  That formatting would be consistent with other tables in the MET/METplus documentation:

https://met.readthedocs.io/en/latest/Users_Guide/reformat_point.html#id5
https://met.readthedocs.io/en/latest/Users_Guide/reformat_point.html#id6
https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements

and I find it easier to read.

File: docs/Users_Guide/release-notes.rst
Comment (georgemccabe) [reply docs/Users_Guide/release-notes.rst]: This command contains fake paths as an example, so it will not work if blindly copy/pasted regardless of where it is run. I could change it to something similar to the above example of a path in the MET installation area:

`<METPLUS_INSTALLATION_LOCATION>/metplus/util/diff_util.py /path/to/output_one /path/to/output_two`

or

`METplus-X.Y.Z/metplus/util/diff_util.py /path/to/output_one /path/to/output_two`

or

`/path/to/METplus-X.Y.Z/metplus/util/diff_util.py /path/to/output_one /path/to/output_two`

What do you think?

Comment (jprestop) [reply docs/Users_Guide/release-notes.rst]: I prefer `/path/to/METplus-X.Y.Z/metplus/util/diff_util.py /path/to/output_one /path/to/output_two`.

File: docs/Users_Guide/release-notes.rst
Comment (jprestop) [reply docs/Users_Guide/release-notes.rst]: I prefer `/path/to/met-X.Y.Z/share/met/GridStatConfig`.

**Comment by georgemccabe** [comment]
@jprestop, to bold the the titles in the first line of those tables, it looks like I would add

```
   :header-rows: 1
```

is that correct? The tables currently have header-rows set to 0

**Comment by jprestop** [comment]
> @jprestop, to bold the the titles in the first line of those tables, it looks like I would add
> 
> ```
>    :header-rows: 1
> ```
> 
> is that correct? The tables currently have header-rows set to 0

I do not the syntax offhand.  You could check out the documentation in the other formatting or the sphinx documentation.  Please let @lisagoodrich or I know if you need assistance.

**Comment by lisagoodrich** [comment]
@jprestop your formatting to bold the first row in a table is correct.
   :header-rows: 1

**Review by jprestop — APPROVED** [review]
@georgemccabe As requested I reviewed the changes to MET Configuration sections in Python Wrappers chapter, reviewed the upgrade instructions, and reviewed the error log formatting on seneca for a use case run that requires updates.  Given the complexity of this guide and these changes, I think it could be useful to have someone who has to make these changes run through this guide.  This would not necessarily need to be before you merge this PR (as I am approving now), but it seems after would be fine as upgrades could be made, in order to obtain feedback on if all of the details are clear and/or if any are missing.  

Thank you for all of this helpful and detailed information. I think this will be valuable to METplus users.

**Pull request merged**
